FGEGraphics/GraphicsHelpers/Models/ShapeGenerators.cs

```diff
@@ -0,0 +1,239 @@
+﻿//
+// This file is part of the Frenetic Game Engine, created by Frenetic LLC.
+// This code is Copyright (C) Frenetic LLC under the terms of a strict license.
+// See README.md or LICENSE.txt in the FreneticGameEngine source root for the contents of the license.
+// If neither of these are available, assume that neither you nor anyone other than the copyright holder
+// hold any right or permission to use this software until such time as the official license is identified.
+//
+
+using FGECore.ModelSystems;
+using System;
+using System.Collections.Generic;
+using OpenTK.Mathematics;
+using FGECore.CoreSystems;
+
+namespace FGEGraphics.GraphicsHelpers.Models;
+
+/// <summary>Generates shapes and returns them as a model.</summary>
+public static class ShapeGenerators
+{
+    /// <summary>
+    /// Generates a 3D sphere model.
+    /// Note that the stacks should be lower than the slices to avoid incorrect normals a good example is 10 stacks and 60 slices.
+    /// </summary>
+    public static Model GenerateSphere(float radius, uint stacks, uint slices, ModelEngine modelEngine)
+    {
+        if (stacks > slices)
+        {
+            Logs.Warning("Sphere has more stacks than slices, this may result in incorrect normals.");
+        }
+        uint vertexCount = stacks * slices;
+        uint numIndices = stacks * slices * 6;
+        List<Vector3> vertices = new((int)vertexCount);
+        List<Vector3> normals = new((int)vertexCount);
+        List<Vector2> texCoords = new((int)vertexCount);
+        uint[] indices = new uint[numIndices];
+        int index = 0;
+        for (uint i = 0; i <= stacks; i++)
```

Comment:
You removed +1 from the vertexCount var but didn't remove the +1 in the actual logic here?

Reply:
What? Where?

Reply:
All through the code. Line 37 has `i = 0; i <= stacks` ie 1 more than stack count, line 55 does `(slices + 1)`, etc. All the code is written to implicitly add 1 to the input values for some reason

Reply:
![bandicam 2024-03-21 11-08-34-779](https://github.com/FreneticLLC/FreneticGameEngine/assets/97306922/171e5503-f0b8-46b6-ad13-a89fccd51a60)

Reply:
I'm gonna take the loss here and say I can't "fix" the more than 1 stack count because I don't think there is an over 1 stack issue. The torus has the + 1 so we can get the end vertex to fully make the torus same concept with cylinder and not have a missing segment in it. The sphere no longer has the + 1 in https://github.com/FreneticLLC/FreneticGameEngine/pull/22/commits/8823665351ba87c6296a2237194c3bec72d33b89

Reply:
The sphere still has the +1 in it

Reply:
What for nextRow? We need that we're making a quad here.

Reply:
`for (uint i = 0; i <= stacks; i++)`
Say `stacks = 3`
this will loop `0, 1, 2, 3` ... this is 4 numbers.
4 is 1 more than 3.
You are calculating for stacks+1.

Reply:
Alright, changed in https://github.com/FreneticLLC/FreneticGameEngine/commit/ecb32d59389d8a2603417f261ea06a8978987d4f it almost worked for cylinder but missing bottom part.

Reply:
Again, I am not referring to _just the specific line I quote_. You need to apply some logical thinking here and process the extended effects of changing the value. It affects every line the variable is in.

Reply:
![bandicam 2024-03-28 19-58-28-315](https://github.com/FreneticLLC/FreneticGameEngine/assets/97306922/d2e17b84-9f91-468f-8779-e0fd88659c9f)

See that missing segment that we need? That's what the + 1 is for to get that extra missing vertex to fully complete the shape.

Reply:
This is going in circles. I don't know what to say at this point other than please turn on the brain and think it through

Reply:
My brain has been turned on and it unfortunately reaches the same conclusion we need that + 1 to fully make the shape I've gone through the code many times. I'm taking the loss and moving on but if you have the correct answer that's been proven not just from a numerical standpoint but visually as well I'd love to see it to learn what I did wrong.

Reply:
Take your views that it won't work as a loss. It is a fact that you don't need the +1. It is your task to make the results correct with a -1 applied properly to normalize such that output shape matches input invariants.